mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -454,40 +426,51 @@ impl PeerManager {
     // If the origin is provided, checks whether this peer is an upstream peer based on configured preferences and
     // connection origin.
     // If the origin is not provided, checks whether this peer is an upstream peer that was seen before.
-    pub fn is_upstream_peer(&self, peer: &PeerNetworkId, origin: Option<ConnectionOrigin>) -> bool {
-        if let Some(origin) = origin {
-            if Self::is_public_downstream(peer.raw_network_id(), origin) {
-                false
-            } else {
-                self.upstream_config
-                    .get_upstream_preference(peer.raw_network_id())
-                    .is_some()
-            }
+    pub fn is_upstream_peer(
+        &self,
+        peer: &PeerNetworkId,
+        metadata: Option<&ConnectionMetadata>,
+    ) -> bool {
+        // Validator network is always upstream
+        if peer.raw_network_id().is_validator_network() {
+            return true;
+        }
+
+        // Outbound connections are upstream on non-P2P networks
+        if let Some(metadata) = metadata {
+            metadata.origin == ConnectionOrigin::Outbound
         } else {
+            // If we already know about the peer, it's upstream
+            // TODO: If this is actually used it seems kinda pointless?
```

Comment:
Yeah, this kinda confused me, too. How would we get into a state where it's been added previously, but it's not good enough to be added now.. 😕

Reply:
I will probably remove this later after some testing, but the only use case that now exists, is that it's a validator which is covered above this.

There is a weird case where the peer is upstream, and we don't have the directional data